db/log_reader.cc

```diff
@@ -450,17 +457,54 @@ unsigned int Reader::ReadPhysicalRecord(Slice* result, size_t* drop_size) {
 
     buffer_.remove_prefix(header_size + length);
 
-    *result = Slice(header + header_size, length);
-    return type;
+    if (uncompress_ && type != kSetCompressionType) {
+      // Uncompress compressed records
+      uncompressed_record_.clear();
+      size_t uncompressed_size = 0;
+      int remaining = 0;
+      do {
+        remaining = uncompress_->Uncompress(header + header_size, length,
+                                            uncompressed_buffer_.get(),
+                                            &uncompressed_size);
+        if (remaining < 0) {
+          buffer_.clear();
+          return kBadRecord;
+        }
+        if (uncompressed_size > 0) {
+          uncompressed_record_.append(uncompressed_buffer_.get(),
+                                      uncompressed_size);
+        }
+      } while (remaining > 0);
```

Comment:
Is ```remaining``` guaranteed to go down to 0 in all cases? I'm still not sure how it works in the case of a chunk compressed by ZSTD_compressStream2() spanning 2 physical records. It seems hard to believe that the uncompression algorithm can completely consume input upto arbitrary boundaries. Would it be safer to concat all the physical records belonging to a logical one and then uncompress?

Reply:
That case works. The uncompression algorithm doesn't rely on knowing the boundaries since the input reader doesn't know them either in case of file compression.

Reply:
Yeah, now that I think about it, I guess in the example I mentioned above the uncompression might keep any input partial data in its internal buffers and flush it to output when it sees the rest. The ZSTD streaming documentation is rather sparse, hence the doubt.

Reply:
If you look at the example here, The outer loop just reads chunks from the compressed file. The inner loop uses the input.pos/input.size to figure out if the chunk can be decompressed any further.

https://github.com/facebook/zstd/blob/dev/examples/streaming_decompression.c#L47